### DISCUSSION TITLE
「Nuxt チュートリアルへようこそ！」の質問

### DISCUSSION THREAD

**Discussion by tokuda109** [vuejs-jp/learn.nuxt.com#85] [ハンズオン Q&A] [answered]
https://github.com/vuejs-jp/learn.nuxt.com/blob/main/content/0.index.md で何か困ったことや質問があれば遠慮なく質問して下さい

**Comment by YuukiIshibashi** [comment]
リアクティビティー パート１にて

> [ref()](https://ja.vuejs.org/api/reactivity-core#ref) は単一の値を保持するためのコンテナを作成し、値が変更された時に自動的に追跡できるようにします。値は .value を通してアクセスすることができます。

とありましたが、`value`以外にもプロパティがあるのでしょうか?

**Reply by kotafujishiro** [reply, marked as answer]
ご質問ありがとうございますー！
利用できるのはvalue以外にはないです。  
詳しくは下記を参照ください。
https://ja.vuejs.org/guide/essentials/reactivity-fundamentals  

（内部的には、、、などあれば詳しいメンバーコメントあればお願いします。

**Reply by ubugeeei** [reply]
あはは，

はい，型定義上は `.value` と，ref であることを型情報に残すためのマーカーだけしか持っていません．

https://github.com/vuejs/core/blob/657603d7b76722b1951a3083e720ab791eed8c28/packages/reactivity/src/ref.ts#L26-L35

https://github.com/vuejs/core/blob/657603d7b76722b1951a3083e720ab791eed8c28/packages/reactivity/src/ref.ts#L55-L61

しかーし！実際には `RefImpl` という internal なクラスで実装されているため，いっぱいプロパティ持ってます．

e.g. `_value`, `_rawValue`, `dep` etc...

https://github.com/vuejs/core/blob/657603d7b76722b1951a3083e720ab791eed8c28/packages/reactivity/src/ref.ts#L108-L159


console.log 等で ref オブジェクトを出力してみると，実際にはいろんなものが詰まってると思います．

**Comment by keishuu** [comment]
配列をリアクティブで管理している場合に、１０列×１０００行ぐらいのデータとなると、
リアクティブの処理が遅くなるのですがそういうものでしょうか？

**Reply by aster-mnch** [reply]
@keishuu 
ご質問ありがとうございます！
そして回答まで日が空いてしまい申し訳ありません 💦 

ref 関数では配列やオブジェクトのネストした値もすべて監視しているため、大規模な配列などには [shallowRef](https://ja.vuejs.org/api/reactivity-advanced#shallowref) 関数を使うことで以下のように `.value` への直接の変更のみを監視することができます。

```ts
const shallowArray = shallowRef([
  /* ディープなオブジェクトの大きなリスト */
])

// これは更新をトリガーせず...
shallowArray.value.push(newObject)
// これはします:
shallowArray.value = [...shallowArray.value, newObject]

// これは更新をトリガーせず...
shallowArray.value[0].foo = 1
// これはします:
shallowArray.value = [
  {
    ...shallowArray.value[0],
    foo: 1
  },
  ...shallowArray.value.slice(1)
]
```

こういった Vue.js を用いたパフォーマンス最適化については、以下の公式ドキュメントが参考になると思いますので合わせてご参照ください。
(今回の大規模配列に関する記述もございますのでぜひ 👍 )

https://ja.vuejs.org/guide/best-practices/performance#general-optimizations

---

なお、こちらの Discussion はハンズオンの Q&A として使用しているため、新たな質問が来ないようにクローズしておりますが、質問対応自体は引き続き行っております。
なので、ご不明点ございましたら引き続きこちらにリプライしていただいても問題ありません 🙇
